dist2src/core.py

```diff
@@ -0,0 +1,406 @@
+        return f"GitRepo(path={self.repo_path}, ref={ref})"
+
+    def checkout(self, branch: str, orphan: bool = False, create_branch: bool = False):
+        """
+        Run `git checkout` in the git repo.
+
+        @param branch: name of the branch to check out
+        @param orphan: Create a branch with disconnected history.
+        @param create_branch: Create branch if it doesn't exist (using -B)
+        """
+        options = {}
+        if orphan:
+            options["orphan"] = branch
+
+        if options:
+            self.repo.git.checkout(**options)
+        elif create_branch:
+            self.repo.git.checkout("-B", branch)
+        else:
+            self.repo.git.checkout(branch)
+
+    def commit(self, message: str):
+        """Commit staged changes in GITDIR."""
+        # some of the commits may be empty and it's not an error, e.g. extra source files
+        self.repo.git.commit("--allow-empty", m=message)
+
+    def commit_all(self, message: str):
+        if self.repo.is_dirty():
+            self.stage()
+            self.commit(message=message)
+        else:
+            logger.info("The repo is not dirty, nothing to commit.")
+
+    def fetch(self, remote: Union[str, Path], refspec: str):
+        """
+        fetch refs from a remote to this repo
+
+        @param remote: str or path of the repo we fetch from
+        @param refspec: see man git-fetch
+        """
+        self.repo.git.fetch(remote, refspec)
+
+    def stage(self, add=None, exclude=None):
+        """ stage content in the repo (git add)"""
+        if exclude:
+            exclude = f":(exclude){exclude}"
+            logger.debug(exclude)
+        self.repo.git.add(add or ".", exclude)
+
+    def create_tag(self, tag, branch):
+        """Create a Git TAG at the tip of BRANCH"""
+        self.repo.create_tag(tag, ref=branch, force=True)
+
+    def cherry_pick_base(self, from_branch, to_branch):
+        """Cherry-pick the first commit of a branch
+
+        Cherry-pick the first commit of FROM_BRANCH to TO_BRANCH in the
+        repository stored in GITDIR.
+        """
+        num_commits = sum(1 for _ in self.repo.iter_commits(from_branch))
+        self.checkout(to_branch, create_branch=True)
+        self.repo.git.cherry_pick(f"{from_branch}~{num_commits - 1}")
+
+
+class Dist2Src:
+    """
+    A convertor for dist-git rpm repos into a source-git variant.
+    """
+
+    def __init__(
+        self,
+        dist_git_path: Optional[Path],
+        source_git_path: Optional[Path],
+        log_level: int = 1,
+    ):
+        """
+        both dist_git_path and source_git_path are optional because not all operations require both
+
+        @param dist_git_path: path to the dist-git repo we want to convert
+        @param source_git_path: path to a source-git repo (doesn't need to exist)
+                                where the conversion output will land
+        @param log_level: int, 0 minimal output, 1 verbose, 2 debug
```

Comment:
Don't we have some enum for that?

Reply:
we don't; this is directly passed from CLI and only used for rpmbuild -vvvvvvvvvvvv

Reply:
I meant `logging.DEBUG`/`logging.INFO`, but they have different numbering if we want to convert that from CLI directly.